dapr/conf/helpers.py

```diff
@@ -1,56 +1,131 @@
-        if len(addr_list[0]) > 0:
-            fqdn = addr_list[0]
-        # Account for Endpoints of the type http://localhost:3500/v1.0/invoke
-        addr_list = addr_list[1].split("/")
-        port = addr_list[0]     # type: ignore
-    elif len(addr_list) == 1:
-        # No port was specified
-        # Account for Endpoints of the type :3500/v1.0/invoke
-        addr_list = addr_list[0].split("/")
-        fqdn = addr_list[0]
-    else:
-        # IPv6 address
-        addr_list = addr.split("]:")
-        if len(addr_list) == 2:
-            # A port was explicitly specified
-            fqdn = addr_list[0]
-            fqdn = fqdn.replace("[", "")
-
-            addr_list = addr_list[1].split("/")
-            port = addr_list[0]     # type: ignore
-        elif len(addr_list) == 1:
-            # No port was specified
-            addr_list = addr_list[0].split("/")
-            fqdn = addr_list[0]
-            fqdn = fqdn.replace("[", "")
-            fqdn = fqdn.replace("]", "")
+from urllib.parse import urlparse, parse_qs
+
+
+class URIParseConfig:
+    DEFAULT_SCHEME = "dns"
+    DEFAULT_HOSTNAME = "localhost"
+    DEFAULT_PORT = 443
+    DEFAULT_TLS = False
+    DEFAULT_AUTHORITY = ""
+    ACCEPTED_SCHEMES = ["dns", "unix", "unix-abstract", "vsock", "http", "https", "grpc", "grpcs"]
+    VALID_SCHEMES = ["dns", "unix", "unix-abstract", "vsock", "grpc", "grpcs"]
+
+
+class GrpcEndpoint:
+    def __init__(self, url: str):
+        self.authority = URIParseConfig.DEFAULT_AUTHORITY
+        self.url = url
+
+        url = self.preprocess_url(url)
+        parsed_url = urlparse(url)
+        validate_path_and_query(parsed_url.path, parsed_url.query, parsed_url.scheme)
+        tls = extract_tls_from_query(parsed_url.query, parsed_url.scheme)
+
+        self.scheme = parsed_url.scheme or URIParseConfig.DEFAULT_SCHEME
+        self.hostname = parsed_url.hostname or URIParseConfig.DEFAULT_HOSTNAME
+        self.port = parsed_url.port or URIParseConfig.DEFAULT_PORT
+        self.tls = tls or URIParseConfig.DEFAULT_TLS
+
+    def is_secure(self) -> bool:
+        return self.tls
+
+    def get_scheme(self) -> str:
+        return self.scheme if self.scheme in URIParseConfig.VALID_SCHEMES \
+            else URIParseConfig.DEFAULT_SCHEME
+
+    def get_port(self) -> str:
+        port = self.get_port_as_int()
+        if port == 0:
+            return ""
+
+        return str(port)
+
+    def get_port_as_int(self) -> int:
+        if self.scheme in ["unix", "unix-abstract"]:
+            return 0
+
+        return self.port
+
+    def get_hostname(self) -> str:
+        hostname = self.hostname
+        if self.hostname.count(":") == 7:
+            # IPv6 address
+            hostname = f"[{hostname}]"
+        return hostname
+
+    def get_endpoint(self) -> str:
+        scheme = self.get_scheme()
+        port = "" if len(self.get_port()) == 0 else f":{self.port}"
+
+        if scheme == "unix":
```

Comment:
Is there a warning if the user uses the deprecated "https" or "http" schemas?

Reply:
Thanks. I added it today.